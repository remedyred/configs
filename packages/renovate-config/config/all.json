{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"extends": [
		"./plugins/*",
		"./automerge/*"
	]
}
