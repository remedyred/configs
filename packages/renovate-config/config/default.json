{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"extends": [
		"./settings/common",
		"./plugins/*",
		"./groups/*",
		"./automerge/automerge",
		"./automerge/automerge-linters",
		"./automerge/automerge-lock",
		"./automerge/automerge-patch",
		"./automerge/automerge-pin",
		"./automerge/automerge-types"
	]
}
