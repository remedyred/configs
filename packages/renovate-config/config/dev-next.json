{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"extends": [
		"./settings/dev-next",
		"./presets/*",
		"./automerge/*",
		"./package-rules/*"
	]
}
