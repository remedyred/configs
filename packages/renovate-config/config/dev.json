{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"extends": [
		"./settings/dev",
		"./plugins/*",
		"./automerge/*",
		"./package-rules/*"
	]
}
