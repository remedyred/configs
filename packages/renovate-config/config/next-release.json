{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"extends": [
		"./settings/next-release"
	]
}
