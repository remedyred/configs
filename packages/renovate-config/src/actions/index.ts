// WARNING: This file is automatically generated. Any changes will be lost the next time the generator is run.

export {default as build} from './build'
export {default as watch} from './watch'
