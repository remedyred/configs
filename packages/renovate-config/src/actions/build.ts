import {$out, Config, SubConfig} from '../common'
import {ParsedArgs} from '@snickbit/node-cli'
import {isEmpty, objectExcept, objectPull, objectSort} from '@snickbit/utilities'
import {getFileJson, saveFileJson} from '@snickbit/node-utilities'
import {validateConfig} from 'renovate/dist/config/validation'
import {mergeChildConfig} from 'renovate/dist/config/utils'
import {RenovateConfig} from 'renovate/dist/config/types'
import {unixify} from 'fast-glob/out/utils/path'
import fg from 'fast-glob'
import path from 'path'

export default async function build(args: ParsedArgs, config: Config) {
	$out.block
		.heading('builder')
		.info(`Building renovate config`)

	return generateFromDefinitionFiles(config)
}

export async function generateFromDefinitionFiles(config: Config): Promise<void> {
	await fromDefinitionFiles('generate', config)
}

export async function buildFromDefinitionFiles(config: Config): Promise<RenovateConfig[]> {
	return fromDefinitionFiles('build', config)
}

async function fromDefinitionFiles(action: 'build' | 'generate', config: Config): Promise<RenovateConfig[]> {
	const definitionFiles = await fg(`${config.source}/*.json`, {absolute: true})

	const promises = []
	for (const definitionFile of definitionFiles) {
		promises.push(action === 'build'
			? buildFromDefinitionFile(definitionFile)
			: generateFromDefinitionFile(definitionFile, config))
	}

	return Promise.all(promises)
}

async function generateFromDefinitionFile(source: string, config: Config): Promise<void> {
	const renovateConfig: RenovateConfig = await buildFromDefinitionFile(source)
	if (!isEmpty(renovateConfig)) {
		let output = `${config?.output || config?.cwd || process.cwd()}/${path.basename(source)}`
		if (!output.endsWith('.json')) {
			output += '.json'
		}
		saveFileJson(output, renovateConfig)
	}
}

export async function buildFromDefinitionFile(source: string): Promise<RenovateConfig> {
	const definition = getFileJson(source)

	const definitionConfig: SubConfig = {
		source,
		fileSources: objectPull(definition, 'extends') || [],
		cwd: path.dirname(source)
	}

	let results: RenovateConfig = {}
	if (definitionConfig.fileSources.length) {
		results = await buildFromDefinition(definitionConfig)
	}
	return results
}

export function cleanSource(fileSource: string, cwd?: string): string {
	let cleanSource = unixify(fileSource).replace(/^\.\//, '').replace(/\/\*$/, '/**/*')
	if (cwd) {
		cleanSource = path.posix.join(cwd, cleanSource)
	}
	if (!cleanSource.endsWith('.json')) {
		cleanSource += '.json'
	}
	return cleanSource
}

export async function buildFromDefinition(config: SubConfig): Promise<RenovateConfig> {
	let combinedFiles: Record<string, unknown> = {}
	const extensions: string[] = []
	const fileSources: string[] = []

	for (const fileSource of config.fileSources) {
		if (fileSource.startsWith('./')) {
			fileSources.push(cleanSource(fileSource, config.cwd))
		} else {
			extensions.push(fileSource)
		}
	}

	const files = await fg(fileSources, {absolute: true})

	for (const file of files) {
		const json: RenovateConfig = getFileJson(file)
		const dirname = path.basename(path.dirname(file))
		const basename = path.basename(file)
		if (json) {
			$out.debug(`adding file ${dirname}/${basename}`)
			if (json.extends) {
				extensions.push(...json.extends)
			}
			combinedFiles = mergeChildConfig(combinedFiles, json)
		}
	}

	combinedFiles = objectSort(objectExcept(combinedFiles, ['extends', '$schema']))

	const results: RenovateConfig = {
		$schema: 'https://docs.renovatebot.com/renovate-schema.json',
		extends: extensions,
		...combinedFiles
	}

	await validate(results)

	return results
}

export async function validate(results: any) {
	const validation = await validateConfig(results, true)

	if (validation.errors.length > 0) {
		$out.fatal('RENOVATE ERROR:', ...validation.errors)
	} else if (validation.warnings.length > 0) {
		$out.warn('RENOVATE WARNING:', ...validation.warnings)
	} else {
		$out.debug('Valid config!')
	}
}
