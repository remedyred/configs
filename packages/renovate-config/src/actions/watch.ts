import {sleep} from '@snickbit/utilities'
import {$out, Config} from '../common'
import {ParsedArgs} from '@snickbit/node-cli'
import chokidar from 'chokidar'
import build from './build'
import debounce from 'debounce'

export default async function watch(args: ParsedArgs, config: Config) {
	const promises: Promise<any>[] = []

	const runBuild = debounce(async () => {
		try {
			await build(args, config)
		} catch (e) {
			$out.error(e.message)
		}

		await sleep(500)
		// remove promise from array
		promises.splice(promises.indexOf(this), 1)

		if (promises.length === 0) {
			$out.info('Watching files for changes...')
		}
	}, 300)

	let ready = false
	chokidar
		.watch([config.source], {
			persistent: true,
			awaitWriteFinish: {
				stabilityThreshold: 2000,
				pollInterval: 100
			}
		})
		.on('ready', () => {
			ready = true
			promises.push(runBuild())
		})
		.on('all', (eventName, path) => {
			if (ready && promises.length < 2) {
				$out.debug('adding build to queue')
				promises.push(runBuild())
			} else {
				$out.verbose({eventName, path})
			}
		})
}
