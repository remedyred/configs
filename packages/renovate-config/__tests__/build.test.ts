import {buildFromDefinitionFile} from '../src/actions/build'
import {validateConfig} from 'renovate/dist/config/validation'
import * as fs from 'fs'
import unixify from 'unixify'
import path from 'path'

const VALID_CONFIG = {errors: [], warnings: []}

const CONFIG_FILE = unixify(path.join(__dirname, '..', 'config', 'dev.json')) // 'config/dev.json'

describe('build.ts', () => {
	it(`config file should exist at ${CONFIG_FILE}`, () => {
		expect(fs.existsSync(CONFIG_FILE)).toBe(true)
	})

	describe('buildFromDefinitionFile', () => {
		it('should return a valid config', async () => {
			const config = await buildFromDefinitionFile(CONFIG_FILE)
			expect(validateConfig(config)).resolves.toMatchObject(VALID_CONFIG)
		})

		it('should return a config with more than 2 keys', async () => {
			const config = await buildFromDefinitionFile(CONFIG_FILE)
			expect(Object.keys(config).length).toBeGreaterThan(2)
		})

		it('should return a config with an extends key with a length greater than 0', async () => {
			const config = await buildFromDefinitionFile(CONFIG_FILE)
			expect(config.extends.length).toBeGreaterThan(0)
		})

		it('should reliably return a valid config', async () => {
			expect(await buildFromDefinitionFile(CONFIG_FILE)).toMatchSnapshot()
		})
	})
})
