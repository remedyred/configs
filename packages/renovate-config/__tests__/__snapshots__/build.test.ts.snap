// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build.ts buildFromDefinitionFile should reliably return a valid config 1`] = `
Object {
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "assignees": Array [
    "snickbit",
  ],
  "automerge": true,
  "automergeStrategy": "squash",
  "automergeType": "pr",
  "baseBranches": Array [
    "dev",
  ],
  "configMigration": true,
  "dependencyDashboard": false,
  "extends": Array [
    "preview:dockerCompose",
    "preview:dockerVersions",
    "regexManagers:dockerfileVersions",
    "workarounds:doNotUpgradeFromAlpineStableToEdge",
    "group:monorepos",
    "group:nodeJs",
    "group:allApollographql",
    "group:fortawesome",
    "group:illuminate",
    "group:jestPlusTSJest",
    "group:jestPlusTypes",
    "group:kubernetes",
    "group:socketio",
    "group:symfony",
    "group:linters",
    "group:jsUnitTest",
    "group:lernaMonorepo",
    "helpers:disableTypesNodeMajor",
    "npm:unpublishSafe",
    "workarounds:typesNodeVersioning",
    "replacements:all",
  ],
  "ignoreDeps": Array [],
  "lockFileMaintenance": Object {
    "automerge": true,
    "automergeType": "pr",
    "enabled": true,
    "platformAutomerge": true,
  },
  "packageRules": Array [
    Object {
      "matchPackagePrefixes": Array [
        "@snickbit",
        "@remedyred",
      ],
      "stabilityDays": 1,
    },
    Object {
      "groupName": "lerna-lite monorepo",
      "matchSourceUrlPrefixes": Array [
        "https://github.com/ghiscoding/lerna-lite",
      ],
      "matchUpdateTypes": Array [
        "digest",
        "patch",
        "minor",
        "major",
      ],
    },
    Object {
      "groupName": "Twilio",
      "matchPackageNames": Array [
        "twilio-*",
        "twilio",
      ],
      "matchPackagePrefixes": Array [
        "@twilio/",
      ],
    },
    Object {
      "automerge": true,
      "matchDepTypes": Array [
        "devDependencies",
      ],
      "matchPackagePatterns": Array [
        "lint",
        "prettier",
        "@snickbit/eslint-config",
      ],
    },
    Object {
      "automerge": true,
      "matchCurrentVersion": "!/^0/",
      "matchUpdateTypes": Array [
        "patch",
      ],
    },
    Object {
      "automerge": true,
      "matchPackagePatterns": Array [
        "^esbuild",
      ],
      "matchUpdateTypes": Array [
        "patch",
      ],
    },
    Object {
      "automerge": true,
      "groupName": "@types packages",
      "matchPackagePatterns": Array [
        "^@types/",
      ],
    },
  ],
  "pin": Object {
    "automerge": true,
  },
  "postUpgradeTasks": Object {
    "commands": Array [
      "pnpm i --lockfile-only",
      "pnpm run fix",
    ],
  },
  "prCreation": "not-pending",
  "rebaseWhen": "conflicted",
  "reviewers": Array [
    "snickbit",
  ],
  "semanticCommits": "enabled",
}
`;
